core/modules/parser/SelectParser.cc

```diff
@@ -118,38 +212,112 @@ class AntlrParser {
         RefAST a = parser.getAST();
     }
 
+    query::SelectStmt::Ptr getStatement() override {
+        if (false == runTransitionDone()) {
+            throw ParseException("Parse error(INTERNAL): run has not been executed on the statement..");
+        }
+        return sf.getStatement();
+        // _selectStmt->diagnose(); // helpful for debugging.
+    }
+
+private:
+    SelectFactory sf;
     std::string statement;
     std::stringstream stream;
     ASTFactory factory;
     SqlSQL2Lexer lexer;
     SqlSQL2Parser parser;
 };
 
+
+class Antlr4Parser : public AntlrParser, public ListenerDebugHelper, public std::enable_shared_from_this<Antlr4Parser> {
```

Comment:
thanks for pinging me about this on slack. it was a good conversation. result is I think it's safe, as long as one does not get the raw pointer out of the shared pointer (generally this is considered to be an unsafe thing to do), and also the factory function with private constructor is worth adding to the class to avoid bad pointer errors trying to use shared from this.